### PR TITLE
Detect mismatch of execution spaces in functors

### LINKED ISSUE #2074: Mismatch between execution_space tag in Functor and execution space in execution policy is silently ignored
This program should have a compilation error:

```c++
using cuda_exec_policy = Kokkos::TeamPolicy<Kokkos::Cuda>;
using serial_exec_policy = Kokkos::TeamPolicy<Kokkos::Serial>;

template <class ExecSpace>
struct MyFunctor {
  using execution_space = ExecSpace;
  KOKKOS_INLINE_FUNCTION
  void operator()(typename serial_exec_policy::member_type const& i) const {
    printf("running Serial version\n");
  }
  KOKKOS_INLINE_FUNCTION
  void operator()(typename cuda_exec_policy::member_type const& i) const {
    printf("running Cuda version\n");
  }
};

int main(int argc , char* argv[]) {

  Kokkos::initialize(argc,argv);

  {
    Kokkos::parallel_for(cuda_exec_policy{1, 1}, MyFunctor<Kokkos::Serial>{});
    Kokkos::fence();

    Kokkos::parallel_for(serial_exec_policy{1, 1}, MyFunctor<Kokkos::Cuda>{});
    Kokkos::fence();
  }

  Kokkos::finalize();

}
```

But it actually just compiles, runs, and prints: 

```
running Cuda version
running Serial version
```

In other words, the functor's `execution_space` attestation is just silently ignored.

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes #2074.

### CODE REVIEW

**Review by ldh4 — APPROVED** [review]
I was in a process of converting my initial solution, similar to this, into a runtime check in attempt to make an unit-test, but I think compile-time check is sufficient, given that it is a fairly simple check.

File: core/src/Kokkos_Parallel.hpp [hunk 1]
```diff
@@ -87,6 +87,31 @@ using device_type_t = typename T::device_type;
...
+  using functor_device_type     = detected_t<device_type_t, Functor>;
+  using functor_device_type_execution_space =
+      detected_t<execution_space_t, functor_device_type>;
+
+  static_assert(
+      std::is_same<policy_execution_space, nonesuch>::value ||
+          std::is_same<functor_execution_space, nonesuch>::value ||
+          std::is_same<policy_execution_space, functor_execution_space>::value,
+      "A policy with an execution space and a functor with an execution space "
+      "are given but the execution space types don't match!");
```
Comment (dalg24) [inline core/src/Kokkos_Parallel.hpp:101]:
```suggestion
      "are given but the execution space types do not match!");
```

**Review by dalg24 — COMMENTED** [review]
I would prefer if you used `is_detected<execution_space_t, Policy>::value` instead of `std::is_same<detected_t<execution_space_t, Policy>, nonesuch>::value`

**Review by cz4rs — APPROVED** [review]

**Comment by masterleinad** [comment]
@dalg24 I addressed your suggestions. 

In my opinion, the changes here would be a good example of why defining `detected_or` would be convenient. I could use the result both for checking that the check was successful and get the referenced type in case of success.

**Review by dalg24 — APPROVED** [review]

**Comment by dalg24** [comment]
> In my opinion, the changes here would be a good example of why defining detected_or would be convenient. I could use the result both for checking that the check was successful and get the referenced type in case of success.

Fair enough.  I still don't like the idea of writing
```C++
using FunctorExecutionSpace = detected_or<MyDefault, execution_space_t, Functor>;
static_assert(!typename FunctorExecutionSpace::value_type::value || ...);
// or
static_assert(!std::is_same<MyDefault, typename FunctorExecutionSpace::type>::value || ...);

**Pull request merged**